

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>pytoda.datasets.drug_affinity_dataset &mdash; pytoda  documentation</title>
  

  
  <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../../index.html" class="icon icon-home"> pytoda
          

          
            
            <img src="../../../_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.1.3
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../api/pytoda.html">API of the pytoda package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">pytoda</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../../index.html">Module code</a> &raquo;</li>
        
      <li>pytoda.datasets.drug_affinity_dataset</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for pytoda.datasets.drug_affinity_dataset</h1><div class="highlight"><pre>
<span></span><span class="sd">&quot;&quot;&quot;Implementation of DrugAffinityDataset.&quot;&quot;&quot;</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">,</span> <span class="n">Tuple</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">torch</span>
<span class="kn">from</span> <span class="nn">torch.utils.data</span> <span class="kn">import</span> <span class="n">Dataset</span>

<span class="kn">from</span> <span class="nn">pytoda.warnings</span> <span class="kn">import</span> <span class="n">device_warning</span>

<span class="kn">from</span> <span class="nn">..proteins.protein_language</span> <span class="kn">import</span> <span class="n">ProteinLanguage</span>
<span class="kn">from</span> <span class="nn">..smiles.smiles_language</span> <span class="kn">import</span> <span class="n">SMILESLanguage</span>
<span class="kn">from</span> <span class="nn">..types</span> <span class="kn">import</span> <span class="n">DrugAffinityData</span>
<span class="kn">from</span> <span class="nn">.protein_sequence_dataset</span> <span class="kn">import</span> <span class="n">ProteinSequenceDataset</span>
<span class="kn">from</span> <span class="nn">.smiles_dataset</span> <span class="kn">import</span> <span class="n">SMILESTokenizerDataset</span>


<div class="viewcode-block" id="DrugAffinityDataset"><a class="viewcode-back" href="../../../api/pytoda.datasets.drug_affinity_dataset.html#pytoda.datasets.drug_affinity_dataset.DrugAffinityDataset">[docs]</a><span class="k">class</span> <span class="nc">DrugAffinityDataset</span><span class="p">(</span><span class="n">Dataset</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Drug affinity dataset implementation.</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="DrugAffinityDataset.__init__"><a class="viewcode-back" href="../../../api/pytoda.datasets.drug_affinity_dataset.html#pytoda.datasets.drug_affinity_dataset.DrugAffinityDataset.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">drug_affinity_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">smi_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">protein_filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">column_names</span><span class="p">:</span> <span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;ligand_name&#39;</span><span class="p">,</span> <span class="s1">&#39;sequence_id&#39;</span><span class="p">,</span> <span class="s1">&#39;label&#39;</span><span class="p">],</span>
        <span class="n">drug_affinity_dtype</span><span class="p">:</span> <span class="n">torch</span><span class="o">.</span><span class="n">dtype</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">int</span><span class="p">,</span>
        <span class="n">smiles_language</span><span class="p">:</span> <span class="n">SMILESLanguage</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">smiles_padding</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">smiles_padding_length</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">smiles_add_start_and_stop</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">smiles_augment</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">smiles_canonical</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">smiles_kekulize</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">smiles_all_bonds_explicit</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">smiles_all_hs_explicit</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">smiles_randomize</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">smiles_remove_bonddir</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">smiles_remove_chirality</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">smiles_vocab_file</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">smiles_selfies</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">smiles_sanitize</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">protein_language</span><span class="p">:</span> <span class="n">ProteinLanguage</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">protein_amino_acid_dict</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;iupac&#39;</span><span class="p">,</span>
        <span class="n">protein_padding</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">protein_padding_length</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">protein_add_start_and_stop</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">protein_augment_by_revert</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">protein_sequence_augment</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="p">{},</span>
        <span class="n">protein_randomize</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
        <span class="n">iterate_dataset</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
        <span class="n">backend</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;eager&#39;</span><span class="p">,</span>
        <span class="n">device</span><span class="p">:</span> <span class="n">torch</span><span class="o">.</span><span class="n">device</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize a drug affinity dataset.</span>

<span class="sd">        Args:</span>
<span class="sd">            drug_affinity_filepath (str): path to drug affinity .csv file. Currently,</span>
<span class="sd">                the only supported format is .csv, with an index and three header</span>
<span class="sd">                columns named as specified in column_names.</span>
<span class="sd">            smi_filepath (str): path to .smi file.</span>
<span class="sd">            protein_filepath (str): path to .smi or .fasta file.</span>
<span class="sd">            column_names (Tuple[str]): Names of columns in data files to retrieve</span>
<span class="sd">                labels, ligands and protein name respectively.</span>
<span class="sd">                Defaults to [&#39;ligand_name&#39;, &#39;sequence_id&#39;, &#39;label&#39;].</span>
<span class="sd">            drug_affinity_dtype (torch.dtype): drug affinity data type.</span>
<span class="sd">                Defaults to torch.int.</span>
<span class="sd">            smiles_language (SMILESLanguage): a smiles language.</span>
<span class="sd">                Defaults to None.</span>
<span class="sd">            smiles_vocab_file (str): Optional .json to load vocabulary. Tries</span>
<span class="sd">                to load metadata if `iterate_dataset` is False.</span>
<span class="sd">                Defaults to None.</span>
<span class="sd">            smiles_padding (bool): pad sequences to longest in the smiles</span>
<span class="sd">                language. Defaults to True.</span>
<span class="sd">            smiles_padding_length (int): manually sets number of applied</span>
<span class="sd">                paddings, applies only if padding is True. Defaults to None.</span>
<span class="sd">            smiles_add_start_and_stop (bool): add start and stop token indexes.</span>
<span class="sd">                Defaults to False.</span>
<span class="sd">            smiles_canonical (bool): performs canonicalization of SMILES (one</span>
<span class="sd">                original string for one molecule), if True, then other</span>
<span class="sd">                transformations (augment etc, see below) do not apply</span>
<span class="sd">            smiles_augment (bool): perform SMILES augmentation. Defaults to</span>
<span class="sd">                False.</span>
<span class="sd">            smiles_kekulize (bool): kekulizes SMILES (implicit aromaticity</span>
<span class="sd">                only). Defaults to False.</span>
<span class="sd">            smiles_all_bonds_explicit (bool): Makes all bonds explicit.</span>
<span class="sd">                Defaults to False, only applies if kekulize = True.</span>
<span class="sd">            smiles_all_hs_explicit (bool): Makes all hydrogens explicit.</span>
<span class="sd">                Defaults to False, only applies if kekulize = True.</span>
<span class="sd">            smiles_randomize (bool): perform a true randomization of SMILES</span>
<span class="sd">                tokens. Defaults to False.</span>
<span class="sd">            smiles_remove_bonddir (bool): Remove directional info of bonds.</span>
<span class="sd">                Defaults to False.</span>
<span class="sd">            smiles_remove_chirality (bool): Remove chirality information.</span>
<span class="sd">                Defaults to False.</span>
<span class="sd">            smiles_selfies (bool): Whether selfies is used instead of smiles.</span>
<span class="sd">                Default to False.</span>
<span class="sd">            smiles_sanitize (bool): RDKit sanitization of the molecule.</span>
<span class="sd">                Defaults to True.</span>
<span class="sd">            protein_language (ProteinLanguage): protein language.</span>
<span class="sd">                Defaults to None, a.k.a., build it from scratch.</span>
<span class="sd">            protein_amino_acid_dict (str): Amino acid dictionary.</span>
<span class="sd">                Defaults to &#39;iupac&#39;.</span>
<span class="sd">            protein_padding (bool): pad sequences to the longest in the protein</span>
<span class="sd">                language. Defaults to True.</span>
<span class="sd">            protein_padding_length (int): manually set the padding.</span>
<span class="sd">                Defaults to None.</span>
<span class="sd">            protein_add_start_and_stop (bool): add start and stop token</span>
<span class="sd">                indexes. Defaults to False.</span>
<span class="sd">            protein_augment_by_revert (bool): augment data by reverting the</span>
<span class="sd">                sequence. Defaults to False.</span>
<span class="sd">            protein_sequence_augment (Dict): a dictionary to specify additional</span>
<span class="sd">                sequence augmentation. Defaults to {}.</span>
<span class="sd">                NOTE: For details please see `ProteinSequenceDataset`.</span>
<span class="sd">            protein_randomize (bool): perform a randomization of the protein</span>
<span class="sd">                sequence tokens. Defaults to False.</span>
<span class="sd">            protein_vocab_file (str): Optional .json to load vocabulary. Tries</span>
<span class="sd">                to load metadata if `iterate_dataset` is False.</span>
<span class="sd">                Defaults to None.</span>
<span class="sd">            iterate_dataset (bool): whether to go through all items in the</span>
<span class="sd">                dataset to extend/build vocab, find longest sequence, and</span>
<span class="sd">                checks the passed padding length if applicable. Defaults to</span>
<span class="sd">                True.</span>
<span class="sd">            backend (str): memory management backend.</span>
<span class="sd">                Defaults to eager, prefer speed over memory consumption.</span>
<span class="sd">                Note that at the moment only theÂ smiles dataset implement both</span>
<span class="sd">                backends. The drug affinity data and the protein dataset are</span>
<span class="sd">                loaded in memory.</span>
<span class="sd">            device (torch.device): DEPRECATED</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">Dataset</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">drug_affinity_filepath</span> <span class="o">=</span> <span class="n">drug_affinity_filepath</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">smi_filepath</span> <span class="o">=</span> <span class="n">smi_filepath</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">protein_filepath</span> <span class="o">=</span> <span class="n">protein_filepath</span>
        <span class="c1"># backend</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">backend</span> <span class="o">=</span> <span class="n">backend</span>
        <span class="n">device_warning</span><span class="p">(</span><span class="n">device</span><span class="p">)</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">column_names</span><span class="p">,</span> <span class="n">Iterable</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Column names was {type(column_names)}, not Iterable.&#39;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">len</span><span class="p">(</span><span class="n">column_names</span><span class="p">)</span> <span class="o">==</span> <span class="mi">3</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Please pass 3 column names not {len(column_names)}&#39;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">column_names</span> <span class="o">=</span> <span class="n">column_names</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">drug_name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">label_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">column_names</span>

        <span class="c1"># SMILES</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">smiles_dataset</span> <span class="o">=</span> <span class="n">SMILESTokenizerDataset</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">smi_filepath</span><span class="p">,</span>
            <span class="n">smiles_language</span><span class="o">=</span><span class="n">smiles_language</span><span class="p">,</span>
            <span class="n">canonical</span><span class="o">=</span><span class="n">smiles_canonical</span><span class="p">,</span>
            <span class="n">augment</span><span class="o">=</span><span class="n">smiles_augment</span><span class="p">,</span>
            <span class="n">kekulize</span><span class="o">=</span><span class="n">smiles_kekulize</span><span class="p">,</span>
            <span class="n">all_bonds_explicit</span><span class="o">=</span><span class="n">smiles_all_bonds_explicit</span><span class="p">,</span>
            <span class="n">all_hs_explicit</span><span class="o">=</span><span class="n">smiles_all_hs_explicit</span><span class="p">,</span>
            <span class="n">remove_bonddir</span><span class="o">=</span><span class="n">smiles_remove_bonddir</span><span class="p">,</span>
            <span class="n">remove_chirality</span><span class="o">=</span><span class="n">smiles_remove_chirality</span><span class="p">,</span>
            <span class="n">selfies</span><span class="o">=</span><span class="n">smiles_selfies</span><span class="p">,</span>
            <span class="n">sanitize</span><span class="o">=</span><span class="n">smiles_sanitize</span><span class="p">,</span>
            <span class="n">padding</span><span class="o">=</span><span class="n">smiles_padding</span><span class="p">,</span>
            <span class="n">padding_length</span><span class="o">=</span><span class="n">smiles_padding_length</span><span class="p">,</span>
            <span class="n">add_start_and_stop</span><span class="o">=</span><span class="n">smiles_add_start_and_stop</span><span class="p">,</span>
            <span class="n">randomize</span><span class="o">=</span><span class="n">smiles_randomize</span><span class="p">,</span>
            <span class="n">vocab_file</span><span class="o">=</span><span class="n">smiles_vocab_file</span><span class="p">,</span>
            <span class="n">iterate_dataset</span><span class="o">=</span><span class="n">iterate_dataset</span><span class="p">,</span>
            <span class="n">backend</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">backend</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># protein sequences</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">protein_sequence_dataset</span> <span class="o">=</span> <span class="n">ProteinSequenceDataset</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">protein_filepath</span><span class="p">,</span>
            <span class="n">protein_language</span><span class="o">=</span><span class="n">protein_language</span><span class="p">,</span>
            <span class="n">amino_acid_dict</span><span class="o">=</span><span class="n">protein_amino_acid_dict</span><span class="p">,</span>
            <span class="n">padding</span><span class="o">=</span><span class="n">protein_padding</span><span class="p">,</span>
            <span class="n">padding_length</span><span class="o">=</span><span class="n">protein_padding_length</span><span class="p">,</span>
            <span class="n">add_start_and_stop</span><span class="o">=</span><span class="n">protein_add_start_and_stop</span><span class="p">,</span>
            <span class="n">augment_by_revert</span><span class="o">=</span><span class="n">protein_augment_by_revert</span><span class="p">,</span>
            <span class="n">sequence_augment</span><span class="o">=</span><span class="n">protein_sequence_augment</span><span class="p">,</span>
            <span class="n">randomize</span><span class="o">=</span><span class="n">protein_randomize</span><span class="p">,</span>
            <span class="n">iterate_dataset</span><span class="o">=</span><span class="n">iterate_dataset</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># drug affinity</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">drug_affinity_dtype</span> <span class="o">=</span> <span class="n">drug_affinity_dtype</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">drug_affinity_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">drug_affinity_filepath</span><span class="p">,</span> <span class="n">index_col</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="c1"># filter data based on the availability</span>
        <span class="n">drug_mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">drug_affinity_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">drug_name</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
            <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">smiles_dataset</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="p">)</span>
        <span class="n">sequence_mask</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">drug_affinity_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">protein_name</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span>
            <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">protein_sequence_dataset</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">drug_affinity_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">drug_affinity_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">drug_mask</span> <span class="o">&amp;</span> <span class="n">sequence_mask</span><span class="p">]</span>
        <span class="c1"># to investigate missing ids per entity</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">masks_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">drug_mask</span><span class="p">,</span> <span class="n">sequence_mask</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">masks_df</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">drug_name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_name</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">number_of_samples</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">drug_affinity_df</span><span class="p">)</span></div>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="s2">&quot;Total number of samples.&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">number_of_samples</span>

    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">index</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DrugAffinityData</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Generates one sample of data.</span>

<span class="sd">        Args:</span>
<span class="sd">            index (int): index of the sample to fetch.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DrugAffinityData: a tuple containing three torch.Tensors,</span>
<span class="sd">                representing respectively: compound token indexes,</span>
<span class="sd">                protein sequence indexes and label for the current sample.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># drug affinity</span>
        <span class="n">selected_sample</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">drug_affinity_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
        <span class="n">affinity_tensor</span> <span class="o">=</span> <span class="n">torch</span><span class="o">.</span><span class="n">tensor</span><span class="p">(</span>
            <span class="p">[</span><span class="n">selected_sample</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">label_name</span><span class="p">]],</span>
            <span class="n">dtype</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">drug_affinity_dtype</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># SMILES</span>
        <span class="n">token_indexes_tensor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">smiles_dataset</span><span class="o">.</span><span class="n">get_item_from_key</span><span class="p">(</span>
            <span class="n">selected_sample</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">drug_name</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="c1"># protein</span>
        <span class="n">protein_sequence_tensor</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">protein_sequence_dataset</span><span class="o">.</span><span class="n">get_item_from_key</span><span class="p">(</span>
            <span class="n">selected_sample</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">protein_name</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">token_indexes_tensor</span><span class="p">,</span> <span class="n">protein_sequence_tensor</span><span class="p">,</span> <span class="n">affinity_tensor</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright PaccMann team.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>