

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>pytoda.datasets.protein_protein_interaction_dataset module &mdash; pytoda  documentation</title>
  

  
  <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
    
    <script type="text/javascript" src="../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="pytoda.datasets.protein_sequence_dataset module" href="pytoda.datasets.protein_sequence_dataset.html" />
    <link rel="prev" title="pytoda.datasets.polymer_dataset module" href="pytoda.datasets.polymer_dataset.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../index.html" class="icon icon-home"> pytoda
          

          
            
            <img src="../_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                0.2.4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="pytoda.html">API of the pytoda package</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="pytoda.data_splitter.html">pytoda.data_splitter module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pytoda.files.html">pytoda.files module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pytoda.transforms.html">pytoda.transforms module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pytoda.types.html">pytoda.types module</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="pytoda.datasets.html">pytoda.datasets package</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.annotated_dataset.html">pytoda.datasets.annotated_dataset module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.base_dataset.html">pytoda.datasets.base_dataset module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.dataframe_dataset.html">pytoda.datasets.dataframe_dataset module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.distributional_dataset.html">pytoda.datasets.distributional_dataset module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.drug_affinity_dataset.html">pytoda.datasets.drug_affinity_dataset module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.drug_sensitivity_dataset.html">pytoda.datasets.drug_sensitivity_dataset module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.gene_expression_dataset.html">pytoda.datasets.gene_expression_dataset module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.polymer_dataset.html">pytoda.datasets.polymer_dataset module</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">pytoda.datasets.protein_protein_interaction_dataset module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#summary">Summary</a></li>
<li class="toctree-l4"><a class="reference internal" href="#reference">Reference</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.protein_sequence_dataset.html">pytoda.datasets.protein_sequence_dataset module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.set_matching_dataset.html">pytoda.datasets.set_matching_dataset module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.smiles_dataset.html">pytoda.datasets.smiles_dataset module</a></li>
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.tests.html">pytoda.datasets.tests package</a></li>
<li class="toctree-l3"><a class="reference internal" href="pytoda.datasets.utils.html">pytoda.datasets.utils package</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="pytoda.preprocessing.html">pytoda.preprocessing package</a></li>
<li class="toctree-l2"><a class="reference internal" href="pytoda.proteins.html">pytoda.proteins package</a></li>
<li class="toctree-l2"><a class="reference internal" href="pytoda.smiles.html">pytoda.smiles package</a></li>
<li class="toctree-l2"><a class="reference internal" href="pytoda.tests.html">pytoda.tests package</a></li>
<li class="toctree-l2"><a class="reference internal" href="pytoda.html#summary">Summary</a></li>
<li class="toctree-l2"><a class="reference internal" href="pytoda.html#reference">Reference</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">pytoda</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="pytoda.html">pytoda package</a> &raquo;</li>
        
          <li><a href="pytoda.datasets.html">pytoda.datasets package</a> &raquo;</li>
        
      <li>pytoda.datasets.protein_protein_interaction_dataset module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/api/pytoda.datasets.protein_protein_interaction_dataset.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-pytoda.datasets.protein_protein_interaction_dataset">
<span id="pytoda-datasets-protein-protein-interaction-dataset-module"></span><h1>pytoda.datasets.protein_protein_interaction_dataset module<a class="headerlink" href="#module-pytoda.datasets.protein_protein_interaction_dataset" title="Permalink to this headline">¶</a></h1>
<p>Implementation of ProteinProteinInteractionDataset.</p>
<div class="section" id="summary">
<h2>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h2>
<p>Classes:</p>
<table class="longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pytoda.datasets.protein_protein_interaction_dataset.ProteinProteinInteractionDataset" title="pytoda.datasets.protein_protein_interaction_dataset.ProteinProteinInteractionDataset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">ProteinProteinInteractionDataset</span></code></a></p></td>
<td><p>PPI Dataset implementation.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="reference">
<h2>Reference<a class="headerlink" href="#reference" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="pytoda.datasets.protein_protein_interaction_dataset.ProteinProteinInteractionDataset">
<em class="property">class </em><code class="sig-name descname">ProteinProteinInteractionDataset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence_filepaths</span></em>, <em class="sig-param"><span class="n">entity_names</span></em>, <em class="sig-param"><span class="n">labels_filepath</span></em>, <em class="sig-param"><span class="n">sequence_filetypes</span><span class="o">=</span><span class="default_value">'infer'</span></em>, <em class="sig-param"><span class="n">annotations_column_names</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">protein_language</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">amino_acid_dict</span><span class="o">=</span><span class="default_value">'iupac'</span></em>, <em class="sig-param"><span class="n">paddings</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">padding_lengths</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">add_start_and_stops</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">augment_by_reverts</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">randomizes</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">device</span><span class="o">=</span><span class="default_value">device(type='cpu')</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytoda/datasets/protein_protein_interaction_dataset.html#ProteinProteinInteractionDataset"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytoda.datasets.protein_protein_interaction_dataset.ProteinProteinInteractionDataset" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Generic</span></code>[<code class="xref py py-obj docutils literal notranslate"><span class="pre">torch.utils.data.dataset.T_co</span></code>]</p>
<p>PPI Dataset implementation. Designed for two sources of protein sequences
and on source of discrete labels.
NOTE: Only supports classification (possibly multitask) but no regression
tasks.</p>
<dl class="py method">
<dt id="pytoda.datasets.protein_protein_interaction_dataset.ProteinProteinInteractionDataset.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sequence_filepaths</span></em>, <em class="sig-param"><span class="n">entity_names</span></em>, <em class="sig-param"><span class="n">labels_filepath</span></em>, <em class="sig-param"><span class="n">sequence_filetypes</span><span class="o">=</span><span class="default_value">'infer'</span></em>, <em class="sig-param"><span class="n">annotations_column_names</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">protein_language</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">amino_acid_dict</span><span class="o">=</span><span class="default_value">'iupac'</span></em>, <em class="sig-param"><span class="n">paddings</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">padding_lengths</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">add_start_and_stops</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">augment_by_reverts</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">randomizes</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">device</span><span class="o">=</span><span class="default_value">device(type='cpu')</span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/pytoda/datasets/protein_protein_interaction_dataset.html#ProteinProteinInteractionDataset.__init__"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#pytoda.datasets.protein_protein_interaction_dataset.ProteinProteinInteractionDataset.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize a protein protein interactiondataset.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sequence_filepaths</strong> (<em>Union</em><em>[</em><em>Files</em><em>, </em><em>Sequence</em><em>[</em><em>Files</em><em>]</em><em>]</em>) – paths to .smi (also as .csv) or .fasta (.gz) file for protein
sequences. For each item in the iterable, one protein sequence
dataset is created. Sequences can be nested, i.e. each protein
sequence dataset can be created from an iterable of filepaths
of same type, see sequence_filetypes.</p></li>
<li><p><strong>entity_names</strong> (<em>Sequence</em><em>[</em><em>str</em><em>]</em>) – List of protein sequence entities,
e.g. [‘Peptides’, ‘T-Cell-Receptors’]. These names should be
column names of the labels_filepaths in order respective to
sequence_filepaths.</p></li>
<li><p><strong>labels_filepath</strong> (<em>str</em>) – path to .csv file with classification
labels.</p></li>
<li><p><strong>sequence_filetypes</strong> (<em>Union</em><em>[</em><em>str</em><em>, </em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em>) – the filetypes of the
sequence files. Can either be a str if all files have identical
types or an Sequence if different entities have different
types. Different types across the same entity are not
supported. Supported formats are {.smi, .csv, .fasta,
.fasta.gz}. Default is <cite>infer</cite>, i.e. filetypes are inferred
automatically.</p></li>
<li><p><strong>annotations_column_names</strong> (<em>Union</em><em>[</em><em>List</em><em>[</em><em>int</em><em>]</em><em>, </em><em>List</em><em>[</em><em>str</em><em>]</em><em>]</em>) – indexes
(positional or strings) for the annotations. Defaults to None,
a.k.a. all the columns, except the entity_names are annotation
labels.</p></li>
<li><p><strong>protein_language</strong> (<a class="reference internal" href="pytoda.proteins.protein_language.html#pytoda.proteins.protein_language.ProteinLanguage" title="pytoda.proteins.protein_language.ProteinLanguage"><em>ProteinLanguage</em></a>) – a ProteinLanguage (or child)
instance, e.g. ProteinFeatureLanguage. Defaults to None,
creating a default instance.</p></li>
<li><p><strong>amino_acid_dict</strong> (<em>str</em>) – The type of amino acid dictionary to map
each sequence token to a unique number. Defaults to ‘iupac’, alternative
is ‘unirep’.</p></li>
<li><p><strong>paddings</strong> (<em>Union</em><em>[</em><em>bool</em><em>, </em><em>Sequence</em><em>[</em><em>bool</em><em>]</em><em>]</em>) – pad sequences to longest in
the protein language. Defaults to True.</p></li>
<li><p><strong>padding_lengths</strong> (<em>Union</em><em>[</em><em>int</em><em>, </em><em>Sequence</em><em>[</em><em>int</em><em>]</em><em>]</em>) – manually sets number
of applied paddings (only if padding = True). Defaults to None.</p></li>
<li><p><strong>add_start_and_stops</strong> (<em>Union</em><em>[</em><em>bool</em><em>, </em><em>Sequence</em><em>[</em><em>bool</em><em>]</em><em>]</em>) – add start and
stop token indexes.  Defaults to False.</p></li>
<li><p><strong>augment_by_reverts</strong> (<em>Union</em><em>[</em><em>bool</em><em>, </em><em>Sequence</em><em>[</em><em>bool</em><em>]</em><em>]</em>) – perform a
stochastic reversion of the amino acid sequence.</p></li>
<li><p><strong>randomizes</strong> (<em>Union</em><em>[</em><em>bool</em><em>, </em><em>Sequence</em><em>[</em><em>bool</em><em>]</em><em>]</em>) – perform a true
randomization of the amino acid sequences. Defaults to False.</p></li>
<li><p><strong>device</strong> (<em>torch.device</em>) – device where the tensors are stored.
Defaults to gpu, if available.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="pytoda.datasets.protein_sequence_dataset.html" class="btn btn-neutral float-right" title="pytoda.datasets.protein_sequence_dataset module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="pytoda.datasets.polymer_dataset.html" class="btn btn-neutral float-left" title="pytoda.datasets.polymer_dataset module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright PaccMann team.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>